### ISSUE TITLE
Consider tile/token embeddings instead of one-hot

### ISSUE THREAD

**Issue by schrum2** [schrum2/MarioDiffusion#37]
This was used in the single-example diffusion example:
https://ojs.aaai.org/index.php/AAAI/article/view/28865

**Comment by Olivia-k18** [comment]
We've still got an issue here, my best guess is that it originates from the get_scene_from_embeddings method in models/general_training_helper.py. In any case, something is still going wrong in getting level scenes back from the diffusion model, despite a pretty small loss. This is something to try to fix before working with tile embeddings, especially when trying to combine levels from different games, or trying to encode the larger tileset of MM-Full.

**Comment by schrum2** [comment]
This will take a lot of discussion before you work on it.

**Comment by fwilson12** [comment]
To test the functionality of the function mentioned below, I trained a block2vec embedding model from scratch, and created a python script that embeds levels in a dataset and then reconstructs them from the embedding model. After running it on the complete Mar1and2_Levels dataset, the method seems to be working as intended:

<img width="1097" height="623" alt="Image" src="https://github.com/user-attachments/assets/42978335-4b0f-420a-863a-3c4cf1d83927" />

> We've still got an issue here, my best guess is that it originates from the get_scene_from_embeddings method in models/general_training_helper.py. In any case, something is still going wrong in getting level scenes back from the diffusion model, despite a pretty small loss. This is something to try to fix before working with tile embeddings, especially when trying to combine levels from different games, or trying to encode the larger tileset of MM-Full.

**Comment by fwilson12** [comment]
Additionally, after training both unconditional and conditional models with the same block2vec embeddings (16 embedding dimensions), the results were noticeably better than the past results in this thread. An interesting thing to note is the difference in quality between the unconditional samples from the purely unconditionally-trained model and the unconditional samples from the text-conditioned model, the latter of which are of much higher quality and coherence

**Pure Unconditional Samples**

- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/5993cc4b-99d1-4cec-8e4a-a0b203a76ced" />
- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/8fe126fa-3b0e-4af4-8fb4-ea2d8ba71b25" />
- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/93009fe9-6b38-4035-bc23-2ed82ad743cb" />
- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/463568cb-c452-4988-a854-648c4f05ed7c" />

**Text-conditioned Unconditional Samples**

- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/0018ff5d-326f-4ea9-a8fa-b0ed76c1fd39" />
- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/34aa03a4-6e39-4010-9ceb-60f26f24f01e" />
- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a52c90f0-47f9-4dc5-acd4-990177aa7884" />
- <img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/2f756671-0559-43b1-a6ad-0a28ad137119" />

**Text-conditioned Samples w/ Caption Input**

- floor with one gap. two platforms. a few enemies. one pipe.
- <img width="371" height="896" alt="Image" src="https://github.com/user-attachments/assets/272943db-b1c5-41e0-86ed-ef55a1eb4a81" />
- <img width="379" height="895" alt="Image" src="https://github.com/user-attachments/assets/18572163-d292-4ff3-a4a3-977662efda2e" />

- full floor. a few platforms. one enemy. one pipe. one cannon.
- <img width="374" height="908" alt="Image" src="https://github.com/user-attachments/assets/a771cfb0-f6fe-4124-801b-6102262894cd" />
- <img width="379" height="960" alt="Image" src="https://github.com/user-attachments/assets/9e3cf764-5f82-4f1f-8ebc-e99155a7ce6c" />

**Comment by fwilson12** [comment]
Next steps are to see if a smaller latent space (~8-10 dims) will produce similarly promising results, working on that next

**Comment by fwilson12** [comment]
Trained an unconditional model with 6 embedding dimensions and the results were pretty horrendous, natural next step is to increase the embedding dimensions until the results improve. It seems that 6 embedding dimensions may not permit the embeddings to be sufficiently distinguishable from one another

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/0e1ae772-8fcd-4890-80d4-62b503f2ab11" />

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/71e2a4cb-09c4-4f1e-a566-194bc9f413ce" />

**Comment by schrum2** [comment]
When a computer is free, or when you have time to set up another work station, please start training a model with embedding vectors of length 10

**Comment by schrum2** [comment]
This is a very confusing and frustrating issue, since the quality of the block embeddings seems to vary a lot, and it is not clear if this is random chance or due to issues with training on different data sets. I just ran the batch file `Mar1and2-unconditional-embeddings.bat` and got fairly bad results:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/080c2e9c-b1b4-4ddb-b416-8b1b485578b5" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/2afdbc1f-734c-4e82-adef-9428c18ef5af" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/c6b013af-ae76-47e9-809d-552191d45b9c" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/c8ab6266-0712-487f-8b97-ccd482860d51" />

This is an unconditional model with block embeddings of length 16 based on data from both SMB1 and SMB2. Fletcher reports good results for a similar model when trained only on data from SMB1. Usually, more data is better. There is a question of where the weak link is. For example, what if we just train a block2vec model on the SMB1 data (or re-use the previously trained model that Fletcher used to get better results), but then train the diffusion model on the combined SMB1/SMB2 data? If that solves the problem, then I guess it would indicate that SMB2 has weird adjacency relationships.

I think that resolving this issue will be fairly important if we want to scale up to bigger domains with more tile types, like Mario Maker 2 #142 so we need to keep working on it.

**Comment by schrum2** [comment]
In the commit above, I modified the model and the training script for block2vec. I'm testing the training now. These changes only exist in by dev_schrum branch. If it works, I'll merge them it, and it not, I'll revert them.

**Comment by schrum2** [comment]
Regarding the modified training script on dev_schrum2: There are both good and bad qualities to the output of the trained model. The good news is that it is not producing any garbled levels with weird or inappropriate block placement. Here are a bunch of good samples:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f9dcb2f0-54e2-4827-aab1-4777035ca940" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/5d1899b4-684d-47f0-ac98-9b035553562c" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a78f890f-ccd4-4bf7-a9dd-02db72d809ac" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a8460c07-84d5-4c2e-aae0-5c76a0aded91" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/ef337f46-b7e3-44fb-bcc3-e3a1fa9e955a" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f7f0858d-2340-46e5-a82a-fbb89e833c0b" />

The bad news is that a high proportion of the generated scenes are completely blank: just blue sky. Also, a lot of the levels that are not completely blank are mostly blank, as in the following:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/0858366d-c6a8-49e0-9635-293ad7504188" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/15142cbf-cf06-4ffc-99c7-6c631dcb52ce" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/3446e308-b557-4b14-94f5-9c9c226df3a5" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/6336e72f-9da1-4188-ac6a-0ded6e276c59" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/87ef423f-cfbe-4d40-8ecb-a5c6d8c3874e" />

In my unconditional samples, 20% were completely blank, and many more were mostly blank. This is clearly a problem, but I'm not sure I want to completely revert all changes since the non-blank outputs look good. Before I try to tweak the training of the block embedding more, I'm going to train a text-conditioned model to see if it does any better. Perhaps the text steering can prevent the blank scenes.

**Comment by schrum2** [comment]
Attempt text-conditioned training with my modified block2vec embeddings looks really terrible at epoch 120 with no indication of improving:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f81d45df-87ff-4619-9b3a-3b4507a2706c" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/7d332e49-c43b-40d7-9607-e83da21000f3" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/bc2af2a2-3453-4063-87a5-bdc4db020e22" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/1f920a40-5bbd-43b4-a06f-7330f4de94bf" />

I'm going to keep messing with my files in dev_schrum2.

@fwilson12 I'm not really sure what guidance to give at this point. We definitely need the block2vec approach to work. You could try focusing on just the SMB1 data. You also have different block2vec training code than I do at this point. We need to figure out what the source of the failure is, and how repeatable it is. If we can create an embedding model that does good job with unconditional diffusion, then we should try re-using that for conditional diffusion.

**Comment by fwilson12** [comment]
Unconditional model trained on mar1and2 data using the 16-dim block2vec tile embedding model trained on SMB1 data produced coherent but mediocre results; notably worse than the unconditional models trained on just SMB1 data with the same block embedding model

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/01be713f-8f4b-4386-9779-d6629ff26ee6" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/f83a24a8-99a9-4147-91b1-02f17d2f1696" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/845d5c2f-08f2-4427-afc1-4906a42a2def" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/4f5363bc-784e-452f-9bf3-0257c483f368" />

**Comment by schrum2** [comment]
Specific things to check:

- [ ] Train a new text-conditional model across Mar1and2 data using the previous SMB1 embedding model (SMB1-OLD)
- [ ] Train a new SMB1 embedding model (SMB1-NEW)
- [ ] Use SMB1-NEW to train new unconditional and conditional models

Previous observations indicate that the previously trained SMB1-OLD was able to lead to unconditional models of reasonable quality on at least two occasions, so we want to see if the same SMB1-OLD model can lead to a good text-conditional model. Assuming that works, we need to figure out if training on SMB1 data is good in general, or if that one good embedding model result was a fluke. That's why we need SMB1-NEW to see if the resuling diffusion models are any good.

All of the above checks need to be conducted by @fwilson12 in his branch. Meanwhile, I'm training with a different version of the code in dev_schrum to see if I can solve our issues by actually changing the training code.

**Comment by fwilson12** [comment]
Using SMB1-OLD, trained another unconditional diffusion model across SMB1 data, results were idential to original run

**Comment by schrum2** [comment]
Recent commit fixed a big problem. Text conditional models were not even using the block embeddings, which explains why none of the text conditional results were any good. Disregard earlier reports about text conditional models and re-train those models (we have two text conditional models training right now).

**Comment by schrum2** [comment]
At home I trained an embedding model on the full Mar1and2 dataset in my dev_schrum branch followed by a text-conditional model and it worked! The main issue really was that stupid error in train diffusion. However, I trained this embedding model with vectors of length 16. Next thing I will try is vectors of length 10.

Although I no longer think that training exclusively on the SMB1 data was the trick to success, I still want @fwilson12 to train a model in his branch to see if the code for training an embedding model works, in contrast with the modified code I have in dev_schrum.

Here are some samples from the text-conditional model I trained with embedding vectors of length 16:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/d352d0b0-181b-44ed-b2dd-f90449c31a02" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/bdb10ac2-1b32-446d-b4ab-12d85cdbfed1" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/ded294ea-8f3a-4c09-afd3-0ff727684f8d" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/bcf3aa43-3ff1-4820-89e4-9fcf66c686ef" />

**Comment by schrum2** [comment]
The conditional diffusion model with length-10 block embeddings also works well using code in dev_schrum:

<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/787d1d37-fc31-492e-bcff-b5bf644bebb1" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/619f4164-8d3d-4f3a-aff3-c6e44502b45d" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/0a12cdac-168f-4fcd-a486-9fde6a349b74" />
<img width="256" height="256" alt="Image" src="https://github.com/user-attachments/assets/a376a70f-6bbf-4996-8e13-7e20ca53ab04" />

This seems to pretty definitively indicate that the block embeddings work, so I guess further testing of the code in the other branch is not really needed. There is still a question of how short the block embedding can be and still work well. I'm going to go ahead and close this issue after all, but I will make new follow-up issues about block embeddings
